### PR TITLE
Fix search engine on complex join; fixes #6397

### LINKED ISSUE #6397: 'Items Seem' search with certain columns displayed causes SQL error
**Describe the bug**

Searching under **Assets > Phones** with **Items seen** filter yields no result.

**Page(s) URL**

/glpi/front/phone.php

**To reproduce**

1. Go to `Assets > Phones`
2. Add `Item seen` search filter (should be default) 
3. Click on `Search`
4. See no result

**Expected behavior**

List of results with fields included on list view matching search keyword 

**Logs**

sql-errors.log
```
glpisqllog.ERROR: DBmysql::query() in /var/www/htdocs/glpi/inc/dbmysql.class.php line 188
  *** MySQL query error:
  SQL: SELECT DISTINCT `glpi_phones`.`id` AS id, 'leoc' AS currentuser,
                        `glpi_phones`.`entities_id`, `glpi_phones`.`is_recursive`,  `glpi_phones`.`name` AS `ITEM_Phone_1`,
                        `glpi_phones`.`id` AS `ITEM_Phone_1_id`,
                        `glpi_phones`.`brand` AS `ITEM_Phone_11`,  `glpi_manufacturers`.`name` AS `ITEM_Phone_23`,  `glpi_states`.`completename` AS `ITEM_Phone_31`,   `glpi_users`.`name` AS `ITEM_Phone_70`,
                        `glpi_users`.`realname` AS `ITEM_Phone_70_realname`,
                        `glpi_users`.`id`  AS `ITEM_Phone_70_id`,
                        `glpi_users`.`firstname` AS `ITEM_Phone_70_firstname`,
                        `glpi_phones`.`contact` AS `ITEM_Phone_7`,  `glpi_phonemodels`.`name` AS `ITEM_Phone_40`,  `glpi_phones`.`serial` AS `ITEM_Phone_5`,   GROUP_CONCAT(DISTINCT CONCAT(IFNULL(`glpi_networkports`.`mac`, '__NULL__'),
                                               '$#$',`glpi_networkports`.`id`) ORDER BY `glpi_networkports`.`id` SEPARATOR '$$##$$')
                              AS `ITEM_Phone_21`,

                  `glpi_locations`.`completename` AS `ITEM_Phone_3`,  `glpi_groups_groups_id_tech`.`completename` AS `ITEM_Phone_49`,  `glpi_phones`.`comment` AS `ITEM_Phone_16`,   GROUP_CONCAT(DISTINCT CONCAT(IFNULL(`glpi_devicefirmwares_c5ad835b6516bc670a3da6ab06d0dcf2`.`version`, '__NULL__'),
                                               '$#$',`glpi_devicefirmwares_c5ad835b6516bc670a3da6ab06d0dcf2`.`id`) ORDER BY `glpi_devicefirmwares_c5ad835b6516bc670a3da6ab06d0dcf2`.`id` SEPARATOR '$$##$$')
                              AS `ITEM_Phone_32` FROM `glpi_phones`LEFT JOIN `glpi_manufacturers`
                                          ON (`glpi_phones`.`manufacturers_id` = `glpi_manufacturers`.`id`
                                              )LEFT JOIN `glpi_states`
                                          ON (`glpi_phones`.`states_id` = `glpi_states`.`id`
                                              )LEFT JOIN `glpi_users`
                                          ON (`glpi_phones`.`users_id` = `glpi_users`.`id`
                                              )LEFT JOIN `glpi_phonemodels`
                                          ON (`glpi_phones`.`phonemodels_id` = `glpi_phonemodels`.`id`
                                              ) LEFT JOIN `glpi_networkports`
                                          ON (`glpi_phones`.`id` = `glpi_networkports`.`items_id`
                                              AND `glpi_networkports`.`itemtype` = 'Phone'
                                              ) LEFT JOIN `glpi_locations`
                                          ON (`glpi_phones`.`locations_id` = `glpi_locations`.`id`
                                              )LEFT JOIN `glpi_groups`  AS `glpi_groups_groups_id_tech`
                                          ON (`glpi_phones`.`groups_id_tech` = `glpi_groups_groups_id_tech`.`id`
                                              ) LEFT JOIN `glpi_items_devicefirmwares`
                                          ON (`glpi_phones`.`id` = `glpi_items_devicefirmwares`.`items_id`
                                              AND `glpi_items_devicefirmwares`.`itemtype` = 'Phone'
                                              ) LEFT JOIN `glpi_devicefirmwares`  AS `glpi_devicefirmwares_c5ad835b6516bc670a3da6ab06d0dcf2`
                                          ON (`glpi_items_devicefirmwares`.`devicefirmwares_id` = `glpi_devicefirmwares_c5ad835b6516bc670a3da6ab06d0dcf2`.`id`
                                              ) WHERE  `glpi_phones`.`is_deleted` = 0  AND `glpi_phones`.`is_template` = 0  AND (  (   (`glpi_phones`.`name`  LIKE '%iphone%'  )  OR  (`glpi_phones`.`brand`  LIKE '%iphone%'  )  OR  (`glpi_manufacturers`.`name`  LIKE '%iphone%'  )  OR  (`glpi_states`.`completename`  LIKE '%iphone%'  ) OR  (((`glpi_users`.`firstname`   LIKE '%iphone%'
                            OR `glpi_users`.`realname`   LIKE '%iphone%'
                            OR `glpi_users`.`name`   LIKE '%iphone%'
                            OR CONCAT(`glpi_users`.`firstname`, ' ', `glpi_users`.`realname`)   LIKE '%iphone%'  )
                            ) )  OR  (`glpi_phones`.`contact`  LIKE '%iphone%'  )  OR  (`glpi_phonemodels`.`name`  LIKE '%iphone%'  )  OR  (`glpi_phones`.`serial`  LIKE '%iphone%'  )  OR  (`glpi_networkports`.`mac`  LIKE '%iphone%'  )  OR  (`glpi_locations`.`completename`  LIKE '%iphone%'  )  OR  (`glpi_groups`.`completename`  LIKE '%iphone%'  )  OR  (`glpi_phones`.`comment`  LIKE '%iphone%'  ))  ) GROUP BY `glpi_phones`.`id` ORDER BY ITEM_Phone_1 ASC
  Error: Unknown column 'glpi_groups.completename' in 'where clause'
  Backtrace :
  inc/search.class.php:1246
  inc/search.class.php:98                            Search::constructData()
  inc/search.class.php:80                            Search::showList()
  front/phone.php:39                                 Search::show()
  {"user":"6@assets.ensemble.com"} 
```

**Screenshots**

**Your GLPI setup (you can find it in Setup > General menu, System tab)**

```
perating system: OpenBSD assets.ensemble.com 6.5 GENERIC.MP#3 amd64
PHP 7.2.17 apache2handler (Core, PDO, Phar, Reflection, SPL, SimpleXML, Zend OPcache, apache2handler, bcmath, calendar, ctype,
	curl, date, dom, exif, fileinfo, filter, ftp, gd, gettext, hash, iconv, imap, json, ldap, libxml, mbstring, mysqli, mysqlnd,
	openssl, pcre, posix, session, sockets, sodium, standard, sysvmsg, sysvsem, sysvshm, tokenizer, wddx, xml, xmlreader, xmlrpc,
	xmlwriter, zlib)
Setup: max_execution_time="600" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
	upload_max_filesize="500M" 
Software: Apache/2.4.39 (Unix) LibreSSL/2.9.1 PHP/7.2.17 ()
	Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36
Server Software: OpenBSD port: mariadb-server-10.0.38p0v1
	Server Version: 10.0.38-MariaDB
	Server SQL Mode: 
	Parameters: assets@localhost/glpidb
	Host info: Localhost via UNIX socket
	
mysqli extension is installed
ctype extension is installed
fileinfo extension is installed
json extension is installed
mbstring extension is installed
iconv extension is installed
zlib extension is installed
curl extension is installed
gd extension is installed
simplexml extension is installed
xml extension is installed
ldap extension is installed
imap extension is installed
Zend OPcache extension is installed
xmlrpc extension is installed
exif extension is installed
APCu extension is not present
CAS extension is not present
Database version seems correct (10.0.38) - Perfect!Database version seems correct (10.0.38) - Perfect!
OK/var/www/htdocs/glpi/files/_log : OK
OK/var/www/htdocs/glpi/config : OK
OK/var/www/htdocs/glpi/files : OK
OK/var/www/htdocs/glpi/files/_dumps : OK
OK/var/www/htdocs/glpi/files/_sessions : OK
OK/var/www/htdocs/glpi/files/_cron : OK
OK/var/www/htdocs/glpi/files/_graphs : OK
OK/var/www/htdocs/glpi/files/_lock : OK
OK/var/www/htdocs/glpi/files/_plugins : OK
OK/var/www/htdocs/glpi/files/_tmp : OK
OK/var/www/htdocs/glpi/files/_cache : OK
OK/var/www/htdocs/glpi/files/_rss : OK
OK/var/www/htdocs/glpi/files/_uploads : OK
OK/var/www/htdocs/glpi/files/_pictures : OK
Web access to the files directory should not be allowed but this cannot be checked automatically on this instance.
Make sure acces to error log file is forbidden; otherwise review .htaccess file and web server configuration.
```

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #6397

### CODE REVIEW

**Comment by cedric-anne** [comment]
Replaces #6447 .

Initial bug was introduced by #6260 and fix #6365 was not enough.

#6388 was required for #6260 to work. The forced "complex join" hack of #6260 is still needed as search engine does not correctly handle join on self without joinparams.
I will not fix this case as I am not confident with any modification on search class.

**Comment by trasher** [comment]
Sounds correct, but tests on addSelect and addWhere are not enough as we've seen. A search must be run, even if if gives no results; just to be sure the SQL is entirely valid.

**Comment by cedric-anne** [comment]
> See my comment about tests

I wanted to change search options tests to run a search for each of them (i.e. using them in criteria) (see #6461), but there are other erros that have to be fixed before being able to do this. So I added tests just on search options present in `providerAddWhere`.

File: tests/functionnal/Search.php [hunk 1]
```diff
@@ -1347,6 +1347,65 @@ public function providerAddWhere() {
...
+                                                 'searchtype' => 'contains',
+                                                 'value'      => 'pc']]];
+      $data = $this->doSearch('Computer', $search_params);
+
+      // check for sql error (data key missing or empty)
+      $this->array($data)
+         ->hasKey('data')
+            ->array['last_errors']->isIdenticalTo([])
+            ->array['data']->isNotEmpty()
+            //expecting no result
```
Comment (trasher) [inline tests/functionnal/Search.php]: wrong comment, could you drop it?

File: tests/functionnal/Search.php [hunk 2]
```diff
@@ -1347,6 +1347,65 @@ public function providerAddWhere() {
...
+      $this->integer((int)$displaypref->add($input))->isGreaterThan(0);
+
+      $data = $this->doSearch('Computer', $search_params);
+
+      // check for sql error (data key missing or empty)
+      $this->array($data)
+         ->hasKey('data')
+            ->array['last_errors']->isIdenticalTo([])
+            ->array['data']->isNotEmpty()
+            //expecting one result
```
Comment (trasher) [inline tests/functionnal/Search.php]: wrong comment, could you drop it?

**Comment by trasher** [comment]
ping @orthagh @btry

**Comment by trasher** [comment]
@btry @tuxxic thanks for the feedback :)

**Pull request merged**